### ISSUE TITLE
Propose a Map Plan signature

### ISSUE THREAD

**Issue by jverkoey** [material-motion/material-motion-experiments-js#92]
The primary goal here is to identify what a reasonable minimal API for a Map plan might look like.
- [x] Pick a website that makes healthy use of mapping as our target.
- [x] Using the material-motion-js runtime v1, create a plan/performer plan that is able to express the mapping behavior shown on the target website.
- [x] Refine the plan type until content with the signature for the purposes of the website.
- [ ] Propose the plan type signature by spec'ing it out in the Starmap.

Note that "Map" may not be the final name for this plan. Feel free to make it more specific to get started faster. E.g. `MapScrollPosition(to:)` would be a much easier domain to start with.

**Comment by jverkoey** [comment]
https://github.com/material-motion/material-motion-experiments-js/issues/90 is related.

**Comment by appsforartists** [comment]
I've been working on a parallax effect:

![parallax](https://cloud.githubusercontent.com/assets/926648/19796092/a24e666a-9c96-11e6-8e98-c1b990e0b6e3.gif)

(This is based on the conversation we had the other day about building a scroll-mapped interaction - I hadn't seen this issue.)  

Does this count as a website making heavy use of mapping, or do you want me to find something in the wild to replicate?

**Comment by appsforartists** [comment]
For the prototype, I have

``` typescript
interface MapScrollPositionToCSSValuePlanArgs {
  propertyName: string;
  valueTemplate: string;
  fromValue: number;
  toValue: number;
  scrollOffset?: number;
  scrollRange?: number;
}
```

in http://codereview.cc/D1817

I'm hesitant to recommend this for the Starmap though.  My hypothesis is that it could be a better abstraction split into steppers and compositions.  
- A composition would be a continuum between 0 and 1 with a current position, e.g. `progress`.  
  
  You can attach other objects to it at any position.  
  
  ``` javascript
  composition.add({ plan: SpringClosed, position: 0, target: element });
  composition.add({ plan: SpringOpen, position: 1 / 3, target: element });
  composition.add({ plan: SpringClosed, position: 2 / 3, target: element });
  ```
  
  When the progress goes from `.32` to `.34`, `SpringOpen` would be emitted; the other direction, `SpringClosed` would be.
  
  You could also attach objects over a range, like an interpolation or another composition:
  
  ``` javascript
  composition.add({ startPosition: 0.5, endPosition: .75, steppable: otherComposition });
  composition.add({ startPosition: 0.25, endPosition: .75, steppable: interpolation });
  ```
  
  Perhaps the interpolation emits a new `planAndTarget` pair each time it receives a new position.  The semantics here are one of the things I will be attacking next.
- The stepper would use an input like time or gesture distance to increment a composition's progress.  A composition could even be a stepper, insofar as it increments the progress of a child composition or interpolation.

That's a summary of my mental model of animation; it's effectively a recursively nestable version of the timeline in {Flash, After Effects, any non-linear editing tool}.  Each node represents a clear mapping from input to output.  (You can see why I was interested in building a reactive version of the Runtime.  I'll be curious to learn Origami Studio; I have a feeling it maps pretty closely to the way I think about this stuff.)

The part that's non-obvious is how it works within the existing runtime.  Perhaps a director creates a composition and adds it to the runtime, using an independent interface from the current `addPlan`?  Since I'm sure you'll want tracability for logging and tooling, the interpolations probably emit plans, which can be caught by a performer that applies the transforms to the target.

Also TBD: how do you make a gestural stepper?  The simplest way would be to use the gestures module I wrote to pull a gesture stream from a target, but that's a tad lower level than `Draggable,Pinchable,Rotatable` (which would make the stream for you).  It's a small difference, but maybe one that's worth ironing out, either now or eventually?

Figuring out how all these pieces connect together with the current system is my next area of work.  That's the bit that feels relevant for the Starmap.  This particular proof-of-concept feels too specific to be worth consideration there.  (Once you have steppers and compositions, you can easily compose `MapScrollPositionToCSSValue` from them.)

**Comment by jverkoey** [comment]
Emitting a new plan for every interpolation change is much too granular: the performer should be handling this directly rather than via the runtime.

**Comment by jverkoey** [comment]
Referring to "Perhaps the interpolation emits a new planAndTarget pair each time it receives a new position." specifically.

Note that none of our other motion libraries have a concept of a plan that changes values directly in a one-off manner. Performers are expected to do this property manipulation themselves. How we share this functionality is a topic for discussion (may be healthy to bake into the runtime), but it is much more likely going to be a function than a plan because we don't expect a tool to have such a granular expression of motion as "set this property value to this".

**Comment by jverkoey** [comment]
> The part that's non-obvious is how it works within the existing runtime. Perhaps a director creates a composition and adds it to the runtime, using an independent interface from the current addPlan? 

It should connect to the runtime like other plans do: the performer is expected to register any listener events and react to them accordingly.

**Comment by jverkoey** [comment]
> Since I'm sure you'll want tracability for logging and tooling, the interpolations probably emit plans, which can be caught by a performer that applies the transforms to the target.

addPlan tracing is for a semantic understanding of what the interaction is doing. Finer-grained real-time tracing will eventually be the responsibility of performers, but is not presently spec'd out.

**Comment by jverkoey** [comment]
> You could also attach objects over a range, like an interpolation or another composition:
> 
> When the progress goes from .32 to .34, SpringOpen would be emitted; the other direction, SpringClosed would be.

I'll emphasize that we need to use real examples to drive need for new APIs before we start investing in them. I haven't found a need for this particular API, do you have an existing production example that shows a clear need?

**Comment by jverkoey** [comment]
> Perhaps the interpolation emits a new planAndTarget pair each time it receives a new position.

Mark brought up a good example of when we might emit plans on each scroll event: when we want to apply a critically-damped SpringTo animation to a css value in reaction to a scroll event. This would cause the value to trail slightly behind rather than be mapped directly to the scroll's value, but could be a preferred solution for performance or user experience considerations.
